src/scaling_of_discrete_log.rs

```diff
@@ -0,0 +1,856 @@
+    >
+    private::PublicParameters<
+        PLAINTEXT_SPACE_SCALAR_LIMBS,
+        ScalarPublicParameters,
+        GroupPublicParameters,
+        PlaintextSpacePublicParameters,
+        RandomnessSpacePublicParameters,
+        CiphertextSpacePublicParameters,
+        EncryptionKeyPublicParameters,
+        CiphertextSpaceValue,
+    >
+where
+    Uint<PLAINTEXT_SPACE_SCALAR_LIMBS>: Encoding,
+{
+    pub fn new<
+        const SCALAR_LIMBS: usize,
+        GroupElement: KnownOrderGroupElement<SCALAR_LIMBS>,
+        EncryptionKey,
+    >(
+        scalar_group_public_parameters: group::PublicParameters<GroupElement::Scalar>,
+        group_public_parameters: GroupPublicParameters,
+        encryption_scheme_public_parameters: EncryptionKeyPublicParameters,
+        ciphertext: CiphertextSpaceValue,
+        upper_bound: Uint<PLAINTEXT_SPACE_SCALAR_LIMBS>,
+    ) -> Self
+    where
+        GroupElement::Scalar: group::GroupElement<PublicParameters = ScalarPublicParameters>,
+        EncryptionKey: AdditivelyHomomorphicEncryptionKey<
+            PLAINTEXT_SPACE_SCALAR_LIMBS,
+            PublicParameters = EncryptionKeyPublicParameters,
+        >,
+        EncryptionKey::PlaintextSpaceGroupElement:
+            group::GroupElement<PublicParameters = PlaintextSpacePublicParameters>,
+        EncryptionKey::RandomnessSpaceGroupElement:
+            group::GroupElement<PublicParameters = RandomnessSpacePublicParameters>,
+        EncryptionKey::CiphertextSpaceGroupElement:
+            group::GroupElement<PublicParameters = CiphertextSpacePublicParameters>,
+    {
+        Self {
+            groups_public_parameters: GroupsPublicParameters {
+                witness_space_public_parameters: (
+                    encryption_scheme_public_parameters
+                        .plaintext_space_public_parameters()
+                        .clone(),
+                    encryption_scheme_public_parameters
+                        .randomness_space_public_parameters()
+                        .clone(),
+                )
+                    .into(),
+                statement_space_public_parameters: (
+                    encryption_scheme_public_parameters
+                        .ciphertext_space_public_parameters()
+                        .clone(),
+                    group_public_parameters,
+                )
+                    .into(),
+            },
+            scalar_group_public_parameters,
+            encryption_scheme_public_parameters,
+            ciphertext,
+            upper_bound,
+        }
+    }
+
+    pub fn group_public_parameters(&self) -> &GroupPublicParameters {
+        let (_, group_public_parameters) = (&self
+            .groups_public_parameters
+            .statement_space_public_parameters)
+            .into();
+
+        group_public_parameters
+    }
+}
+
+pub trait WitnessAccessors<
+    PlaintextSpaceGroupElement: group::GroupElement,
+    RandomnessSpaceGroupElement: group::GroupElement,
```

Comment:
```suggestion
    PlaintextSpaceGroupElement: GroupElement,
    RandomnessSpaceGroupElement: GroupElement,
```
no need to fix prefix everywhere.